### ISSUE TITLE
[Proposal] Improve the docs for `Function.call()`

### ISSUE THREAD

**Issue by navaronbracke** [dart-lang/sdk#48262]
`Function` objects implicitly have a `call()` method.

For instance this is valid code:
```
void main() {
  final Function(int) myFunction = (int value) => print("$value");

  myFunction.call(42); // prints 42 
}
```

### The problem

1) Where is the existence of `call()` documented?
I can't find it in the [docs](https://api.flutter.dev/flutter/dart-core/Function-class.html) for the `Function` class.

2) The Dart plugin suggests the `call()` function when typing. Since the docs don't exist, it cannot provide a link to the implementation. (i.e. blue text with underline in VSCode) This _feels_ a little confusing in my opinion, as if the `Function()` object were dynamic, even though it is not.

Dart SDK version: 2.15.1 (stable) (Tue Dec 14 13:32:21 2021 +0100) on "macos_x64"

**Comment by eernstg** [comment]
Surely this should be described somewhere in the general purpose Dart documentation, but I'm not working on those documents, so I'll leave that to others. 

Note though that it should not be in the documentation of the class `Function`, because there is a difference between `Function` and 'a function type': The latter is an actual function type like `void Function(int)`, but the former is the least common supertype of all function types (so it's a kind of "`dynamic` for functions", and invocations of `call(...)` on a receiver of type `Function` is accepted with no typing constraints).

However, the language specification does specify the required properties:

So, according to the static analysis, a function object has a `call` method whose signature is the same as the function itself, which is specified [here](https://github.com/dart-lang/language/blob/29ef4a3fc1eeb1be3f04804ea1f9cab7c593746f/specification/dartLangSpec.tex#L4942) in the language specification.

The dynamic semantics is specified [here](https://github.com/dart-lang/language/blob/29ef4a3fc1eeb1be3f04804ea1f9cab7c593746f/specification/dartLangSpec.tex#L21326). It says that a function object actually has a `call` method with that signature.

**Comment by lrhn** [comment]
The `Function` type is not an interface type. Not a normal one, at least, so different that it might as well not be one, since it cannot be implemented, and it has a member, `call`, which does not have a function type signature.

The `Function` class declaration is really just a placeholder for the static member `apply` and some documentation.
It doesn't declare the `call` member because it *cannot*. All normal interface members have a function type as their signature. The `call` method on `Function` has `Function` itself as its signature. (That doesn't really make any sense in the normal language semantics, it's special-cased all the way down.)

It's more like `Function` represents a *generalization* over actual function types (like the `dynamic Function(int)` type of `myFunction`). Those have no declarations at all, so there is no place to document them.

So, the `call` member of a function type is defined in the language specification only.

I have no idea why the plugin proposes `call`. I guess it's a *possibility*, but not *necessary*. You can, and should, just write `myFunction(42)`. Writing `myFunction.call(42)` or `myFunction.call.call(42)` might get the same result, but it's unnecessary noise.

Where it might make sense is nullable function references, where writing `maybeFunction?.call(42)` is a way to call in a null-aware way. (Sadly we didn't get the `maybeFunction?(42)` syntax into the language along with `maybeList?[42]`.)

**Comment by navaronbracke** [comment]
@lrhn True, I was using call in a null safe context i.e. `myFunction?.call()`, which is indeed the only use case.
